include/openthread/platform/misc.h

```diff
@@ -105,6 +105,83 @@ void otPlatAssertFail(const char *aFilename, int aLineNumber);
  */
 void otPlatWakeHost(void);
 
+/**
+ * Enumeration of possible NCP power states.
+ *
+ * These are in the same order as the Spinel power state (please see `spinel_power_state_t` and spinel documentation).
+ *
+ * The platform code can decide how to interpret each power state. Here is the high-level expected behavior:
+ *
+ * POWER_STATE_OFF:        NCP is powered off (may be awakened by an external reset).
+ * POWER_STATE_DEEP_SLEEP: NCP may be awakened via an external command or interrupt (or reset).
+ * POWER_STATE_STANDBY:    NCP may be awakened via an internal (e.g., from the radio) event or external trigger.
+ * POWER_STATE_LOW_POWER:  NCP is responsive (and possibly connected), but using less power.
+ * POWER_STATE_ON:         NCP is fully powered.
+ *
+ */
+typedef enum {
+    OT_PLAT_POWER_STATE_OFF        = 0,
+    OT_PLAT_POWER_STATE_DEEP_SLEEP = 1,
+    OT_PLAT_POWER_STATE_STANDBY    = 2,
+    OT_PLAT_POWER_STATE_LOW_POWER  = 3,
+    OT_PLAT_POWER_STATE_ON         = 4,
+} otPlatPowerState;
+
+/**
+ * This function set the desired NCP power state to a given state.
+ *
+ * This is only applicable and used for NCP configuration when `OPENTHREAD_CONFIG_NCP_ENABLE_POWER_STATE_CONTROL`
+ * is enabled.
+ *
+ * @note The NCP is not required to immediately enter the given power-state. The power-state is the desired power state
+ * of NCP when the underlying platform's operating system enters idle mode (i.e., all active tasks/events are processed
+ * and the MCU can potentially enter a low-power state).
+ *
+ * @param[in] aInstance      A pointer to OpenThread instance.
+ * @param[in] aState         The new power state.
+ *
+ * @retval OT_ERROR_NONE     The power state updated successfully.
+ * @retval OT_ERROR_FAILED   The given power state is not supported by the platform.
+ *
+ */
+otError otPlatPowerStateSet(otInstance *aInstance, otPlatPowerState aState);
+
+/**
+ * This function gets the current desired power state.
+ *
+ * This is only applicable and used for NCP configuration when `OPENTHREAD_CONFIG_NCP_ENABLE_POWER_STATE_CONTROL`
+ * is enabled.
+ *
+ * @note The power-state is desired power state of NCP when the underlying platform's operating system enters idle mode
+ * (i.e., all active tasks/events are processed and the MCU can potentially enter low-power state).
+ *
+ * After a reset or NCP wake-up, the power state MUST return `OT_PLAT_POWER_STATE_ON`.
```

Comment:
Is it possible to return `OT_PLAT_POWER_STATE_LOW_POWER`?

Reply:
We can consider this but I like the simplicity of it always being `POWER_STATE_ON`.
Host/user can then choose to set it to `LOW_POWER` if desired (particularly since `LOW_POWER` may be interpreted differently on different platforms).

Reply:
I don't really get the difference between `STATE_ON` and `STATE_LOW_POWER`.  Is this intended to be the way to configure rx-on-when-idle vs. rx-off-when-idle?

In my view, a good implementation will always seek to achieve the lowest power state regardless of the Thread-specific radio configuration (e.g. put CPU into idle or sleep modes).

Reply:
My suggestion is that we practically consider all of the them as enumeration option for power state and leave it to platform on how they are interpreted. 

My opinion/preference is that power state to be only related to NCP's MCU possible power modes and not controlling radio (so not really  rx-on-when-idle vs rx-off-when-idle)... I feel radio should be directly managed by OT core.

Thoughts?

Reply:
The doxygen currently ties them to Thread-specific roles (e.g. router vs. sleepy end device).  At the very least, that needs to change if that's not the intent of this property.

Regarding "platform specific", part of the benefit of spinel is the ability to plug-and-play any OT-based NCP.  My preference to maintain this characteristic.

I agree this property should be related to MCU power modes.  But practically speaking, what is the use case for having `STATE_STANDBY`, `STATE_LOW_POWER`, and `STATE_ON`?  Are there any differences that are externally visible by the host (other than power draw)?

Reply:
To be honest, I don't like the spinel definitions/documentation for this. I was a bit reluctant about changing it, but don't mind simplifying/changing it from this PR.

How about this (to keep it simple and straight-forward):
- We  limit the ot platform power state definitions to `ON` and `SLEEP` 
    - `SLEEP` simply means go to low-power (interpreted by platform code) - we expects platform to notify us when awakened.
   - `ON` means be on and responsive (again platform can decide how to interpret it and even be possibly in different power MCU modes while being `ON`)
- Perhaps keeping `otPlatPowerState` an `enum` (so we can add more options in future (if there is ever a need)).
- Keep the spinel property definitions as is (but change the documentation to avoid confustion)
- Map two of spinel properties `ONLINE` and `DEEP_SLEEP`  to `ON` and `SLEEP` on the `otPlatPowerState` platform APIs (fail on other values).

Thoughts?